### ISSUE TITLE
MongoDBContext as Singleton creates challenges for multi-tenancy scenarios

### ISSUE THREAD

**Issue by dandrejvv** [ffernandolima/mongo-db-data-access#2]
I noticed that the current implementation of `services.AddMongoDbContext` wires up a singleton instance of a given `MongoDbContext`. This assumes that one would never have a connection with a different Mongo DB database (for the likes of multi-tenancy) for the duration of the application lifetime. Is there a reason why it cannot instantiate a new instance per scoped session? (I did notice that the AsyncLocal was causing some trouble when you're invoking it at the wrong stack level and so I was forced to add a workaround of adding a dummy command to the list of commands on creation so as to retain the same list of commands throughout my request).

I have noticed that the current pattern for applying configurations for Mongo Entities is done using the `ApplyConfigurationsFromAssembly` method inside the MongoDbContext constructor. Since its static and not reliant on an instance of a MongoDbContext, one can place that somewhere else (like a static constructor) to be initialized only once (if performance is an issue) especially since it uses the Mongo Bson Class Mapping API.

How would you feel about changing the MongoDb Infrastructure library to cater for this concern so that I can have an out of the box experience that can cater for scenarios where I may have to connect to different Mongo DB instances during an application lifetime?

Thanks

**Comment by ffernandolima** [comment]
Hi @dandrejvv!

According to the [official documentation](https://mongodb.github.io/mongo-csharp-driver/2.18/reference/driver/connecting/#re-use), a `MongoClient` instance should be stored in a global place or registered as `Singleton` in an IoC Container.

As the `MongoDbContext` acts wrapping a `MongoClient` instance internally, it's fair to keep the same lifetime between them.

To sum up, I've done some tests changing the lifetime from `Singleton` to `Transient` and also `Scoped` and the results have not been satisfactory. What happened for instance in a Web API scenario (where each request is executed in a different scope) was a huge amount of live connections breaking the application due to reaching the maximum number of allowed connections.

Talking about the `AsyncLocal` usage, please provide a sample app and some repro steps, so that I can understand what you're exactly talking about in a better way.

Considering that the `MongoDbContext` instance is registered as `Singleton` and also the method `ApplyConfigurationsFromAssembly` should be called only once on its ctor, I don't see any performance issues here. What can be done as a nice-to-have feature is to ensure that the configurations will be applied only once even if the method `ApplyConfigurationsFromAssembly` is called multiple times.

I'm open to discussing enhancements as well as receiving PRs from the community. If you want to draft these enhancements mentioned in this conversation, please feel free to do it.

Thanks.

**Comment by dandrejvv** [comment]
Good day @ffernandolima.
I appreciate your response.
We're interested in submitting a PR for this project to make the library easier to use in database per tenant scenarios. Towards this, I would like to run our proposed changes as well as the rationale for them by you first, to avoid us potentially doing work which you won't accept.

My goal is to have this `MongoDbContext` suitable for use within a multi-tenant environment. That would mean that there would be multiple instances of `MongoClient`, one per tenant. This is to make it possible to connect to different Databases from within the same Application.

In order to achieve this, I would like to propose that we modify the interface of `MongoDbContext` that it would be able to resolve per `Scope` since we have to supply a different connection string to it so that it knows how to access the correct Mongo Database behind the scenes. This will mean having to update the `AddMongoDbUnitOfWork` extension method to register per `Scope` in stead of `Singleton`. It would also mean that the `MongoClient` be encapsulated inside a `ConnectionManager` class that would provide the correct instance of `MongoClient` based on the connection string supplied and ensure it only has one instance per connection string. This will bring things in alignment with how Entity Framework currently operates.
This would also mean that the `_commands` and `_session` inside `MongoDbContext` no longer needs to be `AsyncLocal` since they will be resolved per `Scope`. The `_client` will be abstracted behind a `ConnectionManager` class as mentioned before.

The other concern that I mentioned in my previous message regarding the registration of BsonClassMap instructions could be addressed by having the `AddMongoDbUnitOfWork` extension method provide configuration options to allow the developer to specify which Assembly contains the registration classes. This is similar to how EntityFramework does configuration on Entities. Either this extension method can perform the Assembly scanning to execute the `Configure` methods or we can make the `MongoDbContext` class itself (and not a derived version made by the developer) have a static constructor that performs the scan only once.

I am happy to make those changes if you think that it would be beneficial and then submit a pull request for you to review. Please let me know if there are any other tasks that need to be performed other than developer testing.

Thank you.

**Comment by ffernandolima** [comment]
Hi @dandrejvv!

I believe your suggestions make sense, but I'd like to point out some details that may change the way these enhancements will be designed.

The extension method `AddMongoDbContext` (not the extension method `AddMongoDbUnitOfWork`) is the one that might be changed to receive assembly arguments for the library to scan and map classes.

Also, instead of registering contexts as `Scoped` directly, I'd add an optional argument of type `ServiceLifetime` with the default value `Scoped`, and based on it I'd register the context accordingly. By doing this, the developer would be able to choose the most appropriate lifetime.

To sum up, there are some `AddMongoDbContext` method overloads that mirror the `MongoDbContext` constructors, and I think these methods and also the constructors should be kept due to they are very flexible and help developers to register/resolve context instances easily. It may challenge your changes and the way you will create the `ConnectionManager`. The `ConnectionManager` definition shall have the prefix `MongoDb` to follow the library standard.

The feature which comes from the `AcceptAllChangesOnSave` property must be kept working normally.

Finally, another feature that I drafted [here](https://github.com/ffernandolima/mongo-db-data-access/pull/1) is related to throttle commands and we need to think about how to merge it within these enhancements/changes.

You can undoubtedly submit a PR with these enhancements/changes for me to review. Please adapt the existing unit tests and add new ones if needed. Also, you can attach some artifacts from your tests to the PR description.

Thanks.

**Comment by ffernandolima** [comment]
Hi @dandrejvv!

Is there any news related to this issue?

Thanks.

**Comment by dandrejvv** [comment]
Good day @ffernandolima . 
Apologies for not responding sooner. Our business priorities have changed and so we're unable to give any attention to this right now. If we still decide to continue with this path we'll check in with you again. I hope you were'nt waiting on us for anything related to your existing pull-request.

**Comment by ffernandolima** [comment]
Hi @dandrejv! 
Thanks for answering! Closing this issue for now.
